packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -240,94 +242,20 @@ export class Wallet extends EventEmitter<WalletEvents> {
-   * @returns A promise that resolves when all the objectives are completed
-   */
-  private async ensureObjective(
-    objective: WalletObjective,
-    objectiveMessages: Message[]
-  ): Promise<ObjectiveSuccess | ObjectiveError> {
-    // eslint-disable-next-line no-async-promise-executor
-    return new Promise(async resolve => {
-      try {
-        // If the objective is already done we want to exit immediately
-        if (objective.status === 'succeeded') {
-          this._engine.logger.debug(
-            {objective},
-            'Objective passed into ensureObjective has already succeeded'
-          );
-          resolve({type: 'Success', channelId: objective.data.targetChannelId});
-        }
-        let isComplete = false;
-        this.on('ObjectiveCompleted', (o: WalletObjective) => {
-          if (o.objectiveId == objective.objectiveId && o.status === 'succeeded') {
-            isComplete = true;
-            resolve({type: 'Success', channelId: objective.data.targetChannelId});
-          }
-        });
-        // Now that we're listening for objective success we can now send messages
-        // that might trigger progress on the objective
-
-        await this._messageService.send(objectiveMessages);
-
-        /**
-         * Consult https://github.com/statechannels/statechannels/issues/3518 for background on this retry logic
-         */
-        const {multiple, initialDelay, numberOfAttempts} = this._retryOptions;
-
-        for (let i = 0; i < numberOfAttempts; i++) {
-          if (isComplete) {
-            resolve({channelId: objective.data.targetChannelId, type: 'Success'});
-          }
-          const delayAmount = initialDelay * Math.pow(multiple, i);
-
-          await delay(delayAmount);
-
-          const syncResult = await this._engine.syncObjectives([objective.objectiveId]);
-
-          const messagesForObjective = this.getMessagesForObjective(
-            objective.objectiveId,
-            syncResult
-          );
-
-          await this._messageService.send(messagesForObjective);
-        }
+    const {outbox} = await this._engine.syncObjectives(objectiveIds);
 
-        if (isComplete) {
-          resolve({channelId: objective.data.targetChannelId, type: 'Success'});
-        }
-        resolve({
-          numberOfAttempts: this._retryOptions.numberOfAttempts,
-          type: 'EnsureObjectiveFailed',
-        });
-      } catch (error) {
-        this._engine.logger.error({err: error, objective}, 'Uncaught error in EnsureObjective');
-        resolve({
-          type: 'InternalError' as const,
-          error,
-        });
-      }
+    const results = objectives.map(async o => {
+      const done = this.createObjectiveSuccessPromise(o);
+      return {
+        objectiveId: o.objectiveId,
+        currentStatus: o.status,
+        channelId: o.data.targetChannelId,
+        done,
+      };
     });
+
+    await this.messageService.send(getMessages(outbox));
```

Comment:
You may be sending a LOT of messages at this point. Is the message service expected to do batching?

Reply:
I can imagine some edge cases where that's still an issue (a hub closes shop, closes all virtual channels, wallet crashes, restarts, and jump starts the objectives). But that doesn't seem worth worrying about